vcd/resource_vcd_ui_plugin.go

```diff
@@ -0,0 +1,306 @@
+				Computed:    true,
+				Description: "The UI Plugin vendor name",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The UI Plugin name",
+			},
+			"version": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The version of the UI Plugin",
+			},
+			"license": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The license of the UI Plugin",
+			},
+			"link": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The website of the UI Plugin",
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The description of the UI Plugin",
+			},
+			"status": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The status of the UI Plugin",
+			},
+		},
+	}
+}
+
+func resourceVcdUIPluginCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	uiPlugin, err := vcdClient.AddUIPlugin(d.Get("plugin_path").(string), d.Get("enabled").(bool))
+	if err != nil {
+		return diag.Errorf("could not create the UI Plugin: %s", err)
+	}
+
+	err = publishUIPluginToTenants(vcdClient, uiPlugin, d, "create")
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// We set the ID early so the read function can locate the plugin in VCD, as there's no identifying argument on Create.
+	// All identifying elements such as vendor, plugin name and version are inside the uploaded ZIP file and populated
+	// in Terraform state after a Read.
+	d.SetId(uiPlugin.UIPluginMetadata.ID)
+	return resourceVcdUIPluginRead(ctx, d, meta)
+}
+
+// publishUIPluginToTenants performs a publish/unpublish operation for the given UI plugin.
+func publishUIPluginToTenants(vcdClient *VCDClient, uiPlugin *govcd.UIPlugin, d *schema.ResourceData, operation string) error {
+	if d.HasChange("tenant_ids") {
+		orgsToPublish := d.Get("tenant_ids").(*schema.Set).List()
+		existingOrgs, err := vcdClient.GetOrgList()
+		if err != nil {
+			return fmt.Errorf("could not publish the UI Plugin %s to Organizations '%v': %s", uiPlugin.UIPluginMetadata.ID, orgsToPublish, err)
+		}
+		var orgsToPubRefs types.OpenApiReferences
+		for _, org := range existingOrgs.Org {
+			for _, orgId := range orgsToPublish {
+				// We do this as org.ID is empty, so we need to reconstruct the URN with the HREF
+				uuid := extractUuid(orgId.(string))
+				if strings.Contains(org.HREF, uuid) {
+					orgsToPubRefs = append(orgsToPubRefs, types.OpenApiReference{ID: "urn:cloud:org:" + uuid, Name: org.Name})
+				}
+			}
+		}
+		if operation == "update" {
+			err = uiPlugin.UnpublishAll() // We need to clean up the already-published Orgs to put the new ones during an Update.
+			if err != nil {
+				return fmt.Errorf("could not publish the UI Plugin %s to Organizations '%v': %s", uiPlugin.UIPluginMetadata.ID, orgsToPublish, err)
+			}
+		}
```

Comment:
Will this not cause interruptions to tenants? As I read it - simply adding a new tenant will cause a temporary loss of plugin access for the ones that already have it (not sure if in some cases it can cause tenant configuration loss of any sort).

I am concerned this might impact bigger deployments (especially if the `publish` operation fails for whatever reason in the middle). 

Could we try to explicitly unpublish for the ones that should lose access (something similar we have done here https://github.com/vmware/terraform-provider-vcd/blob/main/vcd/resource_vcd_catalog_item.go#L252-L277)

Reply:
Also a point to consider: could there be plugins that need to be active all the time and if unavailable, need to be reconfigured?

Reply:
@Didainius I pushed the change.

Reply:
> @Didainius I pushed the change. I changed a bit the whole function, and also added some comments to understand why is done like that.

This worked well in my manual testing. Thanks for fixes.

Reply:
Feel free to close this thread unless @adezxc has any more queries.